[Inductor] Build FX Linear + Permute Vertical Fusion in Inductor

jiawenliu64 opened the pull request:
Summary:
Build fx-based linear/matmul/bmm + permute/transpose vertical fusion in Inductor

For an internal Ads model: 1.15x -> 1.36x speedup

Differential Revision: D41071665

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -340,6 +340,7 @@ def compile_fx(model_: torch.fx.GraphModule, example_inputs_: List[torch.Tensor]
         model_ = normalize_ir(model_, example_inputs_)
         model_ = overrides.replace_fx(model_)
         model_ = overrides.fuse_fx(model_, example_inputs_)
+        model_ = overrides.fxFuser(model_, example_inputs_)
```
This seems to be doing a lot of the same things as the fuse_fx line right above it (which is CPU-only) is doing.  There are also a few pending PRs there (see this PR stack: https://github.com/pytorch/pytorch/pull/87064)

Can we refactor `fuse_fx` and `fxFuser` to share more code?

cc @XiaobingSuper

jiawenliu64 replied:
Yes, we could integrate fusion of `fxFuser` into `fuse_fx`. I will place the fusion of `fxFuser` before `if torch.is_grad_enabled():', because it supports both inference and training and not is limited to CPU.

jgong5 commented on torch/_inductor/overrides.py:
```diff
@@ -396,6 +406,195 @@ def _philox_rand_like(input, seed, offset):
+# Y1 = X * W^T + bias
+# Y2 = Y1.permute(0, 2, 1)
+# ---->
+# Y2 = (W * X^T + bias.unsqueeze(-1))^T
```
@jiawenliu64 May I know why the FX-level transformation can bring perf benefit? Inductor should handle the transposition/permute views properly without additional copies. So the compute cost should be the same? Same question for other transformations in this PR.

jiawenliu64 replied:
Hi @jgong5, the linear + permute transformations are performed very well in Inductor. As Jason pointed out, it's related to the heuristic in Inductor. Our next step is to optimize the heuristic in Inductor codegen and would remove this FX optimization in the future

jgong5 replied:
Thanks for the reply @jiawenliu64 . May I know what particular cases caused difference in codegen after you applied the transformations? The reason I'm asking is because I tried two cases mentioned in your comments but the generated code seems the same with the pre and post transformation, at least with cpp codegen.
```python
# Y1 = X * W^T + bias
# Y2 = Y1.permute(0, 2, 1)
# ---->
# Y2 = (W * X^T + bias.unsqueeze(-1))^T
```
```python
# X1 = X.permute(0, 2, 1)
# Y1 = X1 * W1^T + bias1
# ---->
# Y2 = X1.transpose(-1, -2) * W1^T + bias1
```

jansel reviewed (approved):
I feel like this might be pointing to a more underlying bug in the way inductor handles layouts for matmuls.  I would not expect this change to matter for performance.

jianyuh commented on torch/_inductor/overrides.py:
```diff
@@ -396,6 +406,195 @@ def _philox_rand_like(input, seed, offset):
+def linear_permute_fusion(module: torch.fx.GraphModule) -> torch.fx.GraphModule:
+    for node in module.graph.nodes:
+        if (
+            node.op == "call_method"
```
Curious if we can make this more general to support both "call_method" as well as "call_function"? Like people can use `tensorX.permute(0, 2, 1)`, or `torch.permute(tensorX, (0, 2, 1))`.

jiawenliu64 commented:
@pytorchmergebot merge

Commit:
Build FX Linear + Permute Vertical Fusion in Inductor (#88566)

Summary:
Build fx-based linear/matmul/bmm + permute/transpose vertical fusion in Inductor

For an internal Ads model: **1.15x -> 1.36x speedup**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/88566

Test Plan: CI

Reviewed By: bertmaher, jansel, jianyuh

Differential Revision: D41071665

Pulled By: jiawenliu64

fbshipit-source-id: 25efb75c83f6d965593f1fe20a5c6a315ccf5ca8

jiawenliu64 commented:
@pytorchmergebot merge -f “macos test failed also on the trunk"

jianyuh commented:
@pytorchbot --help

jiawenliu64 commented:
@pytorchbot merge -f ‘macos test failed also on the trunk’

jiawenliu64 commented:
@pytorchbot rebase -s

jiawenliu64 commented:
@pytorchbot merge -f 'all failures are also on the trunk'

huydhn commented:
@pytorchbot revert -m 'This change breaks trunk https://hud.pytorch.org/pytorch/pytorch/commit/48b58930cbfa725ac25a9303d496c76bf983574d' -c ignoredsignal
